modules/avatar/react/lib/Avatar.tsx

```diff
@@ -10,80 +11,136 @@ export enum AvatarVariant {
   /**
    * The variant of the Avatar default state. Accepts `Light` or `Dark`.
    * @default AvatarVariant.Light
    */
-  variant: AvatarVariant;
+  variant?: AvatarVariant;
   /**
    * The size of the Avatar.
    * @default SystemIconCircleSize.m
    */
-  size: SystemIconCircleSize | number;
+  size?: SystemIconCircleSize | number;
   /**
    * The alt text of the Avatar image.
    * @default Avatar
    */
-  altText: string;
+  altText?: string;
   /**
    * The url of the Avatar image.
    */
   url?: string;
+  /**
+   * The alternative container type for the button. Uses Emotion's special `as` prop.
+   * Will render an `button` tag instead of a `div` when defined.
+   */
+  as?: 'div';
+  /**
+   * The function called when the Avatar is clicked.
+   */
+  onClick?: React.MouseEventHandler<HTMLButtonElement>;
+  /**
+   * The ref to the button that the styled component renders.
+   */
+  buttonRef?: React.Ref<HTMLButtonElement>;
 }
 
-export interface AvatarProps extends AvatarLocalProps, React.HTMLAttributes<HTMLDivElement> {}
-
-export const avatarStyles: CSSObject = {
-  background: colors.soap200,
-  display: 'flex',
-  alignItems: 'center',
-  justifyContent: 'center',
-  padding: 0,
-  border: 0,
-  borderRadius: borderRadius.circle,
-  boxSizing: 'border-box',
-  overflow: 'hidden',
-  '& img': {
-    width: '100%',
-    height: '100%',
+const StyledContainer = styled('button', {
+  shouldForwardProp: prop => isPropValid(prop) && prop !== 'size',
+})<Pick<AvatarProps, 'variant' | 'size' | 'onClick'>>(
+  {
+    background: colors.soap200,
+    position: 'relative',
+    display: 'flex',
+    alignItems: 'center',
+    justifyContent: 'center',
+    padding: 0,
+    border: 0,
+    boxSizing: 'border-box',
+    overflow: 'hidden',
+    borderRadius: borderRadius.circle,
   },
-};
+  ({size}) => ({
+    height: size,
+    width: size,
+  }),
+  ({variant, onClick}) => ({
+    cursor: onClick ? 'pointer' : 'default',
+    '&:not([disabled])': {
+      '&:focus': {
+        outline: 'none',
+        ...(variant === AvatarVariant.Dark ? focusRing(2, 2) : focusRing(2)),
+      },
+    },
+    ...hideMouseFocus,
+  })
+);
 
-const StyledContainer = styled('div', {
-  shouldForwardProp: prop => isPropValid(prop) && prop !== 'size',
-})<Pick<AvatarProps, 'variant' | 'size'>>(
+const StyledStack = styled('span')<Pick<AvatarProps, 'size'>>(
   {
-    ...avatarStyles,
+    position: 'absolute',
   },
   ({size}) => ({
     height: size,
     width: size,
   })
 );
 
-export default class Avatar extends React.Component<AvatarProps> {
-  static Variant = AvatarVariant;
-  static Size = SystemIconCircleSize;
+const StyledImage = styled('img')<{isLoaded: boolean}>(
+  {
+    width: '100%',
+    height: '100%',
+    borderRadius: borderRadius.circle,
+    transition: 'opacity 200ms linear',
+  },
+  ({isLoaded}) => ({
+    opacity: isLoaded ? 1 : 0,
+  })
+);
+
+const Avatar = ({
```

Comment:
Am I missing the function overloading we discussed? If it's not included, when I pass `as="div"`, all of the button attributes will still be valid which is not what we want

Reply:
Sorry I don't remember that conversation? Are you saying we should null out onClick if as='div' is passed in?

Reply:
Maybe my initial explanation wasn't clear. Check out https://github.com/Workday/canvas-kit/blob/prerelease/v4/modules/button/react/lib/DropdownButton.tsx#L33 and https://github.com/Workday/canvas-kit/blob/prerelease/v4/modules/button/react/lib/types.ts#L112.

`ButtonOrAnchorComponent` provides different interfaces depending on whether `as` is defined by using function overrides. We'll probably want to do something similar here. Since it's only in one place and not shared a bunch of components (like buttons), you could probably just put the type inline. Something like

```tsx
type AvatarDivProps<P> = Omit<
  P,
  keyof React.ButtonHTMLAttributes<HTMLButtonElement>
> &
  React.DivHTMLAttributes<DivAnchorElement>;

type AvatarOverload = {
(props: AvatarProps): React.ReactElement;
(props: {as: 'div'} & AvatarDivProps<ButtonProps>): React.ReactElement;
  Variant: typeof AvatarVariant;
  Size: typeof AvatarSize;
};
const Avatar: AvatarOverload = ({
//...
```

We'll likely want a more generic type for this in the future, but I wouldn't worry about it in this PR.

Reply:
done, one thing I noticed is I had to  `Omit<AvatarProps, 'as'>` from the button version otherwise ts wouldn't complain about adding ButtonHTMLAttributes to as='div' Avatars.